ibis/backends/flink/__init__.py

```diff
@@ -853,3 +859,142 @@ def insert(
             "is not a pandas DataFrame or is not a ibis Table."
             f"The given obj is of type {type(obj).__name__} ."
         )
+
+    def to_pyarrow(
+        self,
+        expr: ir.Expr,
+        *,
+        params: Mapping[ir.Scalar, Any] | None = None,
+        limit: int | str | None = None,
+        **kwargs: Any,
+    ) -> pa.Table:
+        from ibis.formats.pyarrow import PyArrowSchema
+
+        pyarrow_batches = self.to_pyarrow_batches(
+            expr, params=params, limit=limit, **kwargs
+        )
+
+        try:
+            pa_table = pa.Table.from_batches(pyarrow_batches)
+
+        except ValueError:
+            table_expr = expr.as_table()
+            ibis_schema = table_expr.schema()
+            arrow_schema = PyArrowSchema.from_ibis(ibis_schema)
+            pa_table = pa.Table.from_batches(pyarrow_batches, schema=arrow_schema)
+
+        return expr.__pyarrow_result__(pa_table)
+
+    def to_pyarrow_batches(
+        self,
+        expr: ir.Table,
+        *,
+        params: Mapping[ir.Scalar, Any] | None = None,
+        chunk_size: int | None = None,
+        limit: int | str | None = None,
+        **kwargs: Any,
+    ):
+        try:
+            return self.to_pyarrow_batches_w_batch_iterator(
+                expr=expr,
+                params=params,
+                chunk_size=chunk_size,
+                limit=limit,
+                **kwargs,
+            )
+
+        except ValueError:
```

Comment:
Similar question here: can we check the conditions that would trigger the error instead of catching the error?

Reply:
Thanks. Split the logic into smaller functions and now checking the conditions first instead of catching the error.